Forward port changes from `humble` to `master` branch

mossmaurice opened the pull request:
Cherry-picked the following PRs:

* https://github.com/ros2/rmw_iceoryx/pull/84 :heavy_check_mark: 
* https://github.com/ros2/rmw_iceoryx/pull/90 :heavy_check_mark: 
* https://github.com/ros2/rmw_iceoryx/pull/97 :heavy_check_mark: 
* https://github.com/ros2/rmw_iceoryx/pull/96 :heavy_check_mark:

elBoberido commented on rmw_iceoryx_cpp/src/types/iceoryx_client.hpp:
```diff
@@ -0,0 +1,43 @@
+    request_size_(rmw_iceoryx_cpp::iceoryx_get_request_size(type_supports))
+  {}
+
+  rosidl_service_type_support_t type_supports_;
+  iox::popo::UntypedClient * const iceoryx_client_;
+  bool is_fixed_size_;
+  size_t request_size_;
+  int64_t sequence_id_;
```
The chicken in me would like to initialize all of this :)
There is no performance impact on the members which are initialized in the ctor and in case a new ctor will be added in the future and one forgets to initialize all members, no undefined behavior will be introduced. Saying that, `sequence_id_` should be initialized to something meaningful. Compiler are [weird beasts](https://devblogs.microsoft.com/oldnewthing/20140627-00/?p=633) when doing optimizations

mossmaurice replied:
@elBoberido Good point, fixed.

elBoberido commented on rmw_iceoryx_cpp/src/types/iceoryx_server.hpp:
```diff
@@ -0,0 +1,47 @@
+    response_size_(rmw_iceoryx_cpp::iceoryx_get_response_size(type_supports))
+  {}
+
+  rosidl_service_type_support_t type_supports_;
+  iox::popo::UntypedServer * const iceoryx_server_;
+  bool is_fixed_size_;
+  size_t response_size_;
```
Same here. Our future selves will be grateful :)

mossmaurice replied:
@elBoberido Good point, fixed.

elBoberido reviewed (commented):
I just had a look at the code in general and iceoryx specifics in dept. For ROS specific changes it's better when someone does the reviews instead of relying on my ignorance :)

mossmaurice commented on rmw_iceoryx_cpp/src/internal/iceoryx_deserialize.cpp:
```diff
@@ -50,4 +51,40 @@ void deserialize(
+  const rosidl_service_type_support_t * type_supports,
+  void * ros_message)
+{
+  auto ts = get_type_support(type_supports);
+
+  if (ts.first == TypeSupportLanguage::CPP) {
```
@elBoberido Yes, the underlying ROS typesupport API [handles](https://docs.ros.org/en/rolling/p/rosidl_runtime_c/generated/function_message__type__support__struct_8h_1aedbefa8782258a0c3184eb05fa49eb7f.html) a `nullptr`, see [here](https://github.com/ros2/rmw_iceoryx/blob/humble/rmw_iceoryx_cpp/src/internal/iceoryx_type_info_introspection.cpp#L144-L164).

mossmaurice commented on rmw_iceoryx_cpp/src/internal/iceoryx_deserialize_typesupport_c.hpp:
```diff
@@ -239,6 +241,22 @@ const char * deserialize(
+  const rosidl_typesupport_introspection_c__ServiceMembers * service_members,
+  void * ros_message)
+{
+  return deserialize(serialized_msg, service_members->response_members_, ros_message);
```
@elBoberido This is also handled by `get_message_typesupport_handle`, [see above thread](https://github.com/ros2/rmw_iceoryx/pull/98#discussion_r1328412533).

mossmaurice commented on rmw_iceoryx_cpp/src/internal/iceoryx_name_conversion.cpp:
```diff
@@ -186,4 +191,30 @@ get_iceoryx_service_description(
+  std::string package_name;
+  std::string type_name;
+  extract_type(type_supports, package_name, type_name);
+  type_name = package_name + "/" + type_name;
```
@elBoberido Good point, fixed.

mossmaurice replied:
@elBoberido There's no difference, but I was already confused when I appended to `package_name` and then passed it to `make_service_description` as `type_name`. Something for a follow-up PR.

mossmaurice commented on rmw_iceoryx_cpp/src/internal/iceoryx_topic_names_and_types.cpp:
```diff
@@ -140,6 +143,46 @@ std::map<std::string, std::string> get_topic_names_and_types()
+  // Can be discovered, but missing associated node
+  std::map<std::string, std::vector<std::string>> topic_servers;
+
+  std::vector<iox::capro::ServiceDescription> available_servers;
+
+  iox::runtime::ServiceDiscovery service_discovery;
+  service_discovery.findService(
+    iox::cxx::nullopt,
+    iox::cxx::nullopt,
+    iox::cxx::nullopt,
+    [&](auto & service_description) {available_servers.push_back(service_description);},
+    iox::popo::MessagingPattern::REQ_RES);
```
@elBoberido That is a trade-off in terms of wasting memory vs. reallocation time. If the capacity is not large enough `std::vector` uses the strategy to allocate twice the old capacity on `push_back`, hence this should be a sufficient trade-off here.

elBoberido replied:
I was also thinking about the time/memory tradeoff but after estimating the memory I had a tendency towards my proposal. I doesn't matter that much anyway since this should not be in a performance critical path :)

mossmaurice commented on rmw_iceoryx_cpp/src/internal/iceoryx_type_info_introspection.cpp:
```diff
@@ -218,6 +313,23 @@ std::string iceoryx_get_message_name(const rosidl_message_type_support_t * type_
+    return members->service_name_;
+  }
+  // Something went wrong
+  return "";
```
@elBoberido I understand what you mean. However, `std::unreachable` would be wrong as there might be other language type supports than C or C++. Maybe the comment is misleading here. Any error with the extraction of the type support will be reported by `get_{service,message}_typesupport_handle`. If you don't mind I would keep this unchanged for now and tackled it with a wider refactoring.

elBoberido replied:
Sure. It's up to you. This just comes from the philosophy to fail fast instead of hunting down a bug :)

The same pattern was also used at different places so it should be done in a consistent way and if changed, it should therefore be changed everywhere which would blow up this PR.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_client.cpp:
```diff
@@ -34,8 +38,94 @@ rmw_create_client(
+  rmw_client = rmw_client_allocate();
+  if (!rmw_client) {
+    RMW_SET_ERROR_MSG("failed to allocate memory for client");
+    cleanupAfterError();
+    return nullptr;
```
@elBoberido Hehe, thanks :grin:

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_request.cpp:
```diff
@@ -47,7 +110,83 @@ rmw_take_request(
+  iceoryx_server->take()
+  .and_then(
+    [&](const void * iceoryx_request_payload) {
+      const iox::mepoo::ChunkHeader * chunk_header =
```
Thanks fixed.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_response.cpp:
```diff
@@ -17,38 +18,178 @@
+      auto iceoryx_response_header = iox::popo::ResponseHeader::fromPayload(
+        iceoryx_response_payload);
+
+      if (iceoryx_response_header->getSequenceId() ==
+      iceoryx_client_abstraction->sequence_id_ - 1)
```
@elBoberido Correct, that is a known limitation currently, [see here](https://github.com/ros2/rmw_iceoryx/releases/tag/v2.0.3-humble). I have planned to open a API change proposal in [rmw](https://github.com/ros2/rmw) to add the `void*` to ` rmw_request_id_t`.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_response.cpp:
```diff
@@ -17,38 +18,178 @@
+          ret = RMW_RET_ERROR;
+          return;
+        }
+        memcpy(request_header->request_id.writer_guid, &guid, size);
+        request_header->request_id.sequence_number = iceoryx_response_header->getSequenceId();
+        request_header->source_timestamp = 0;  // Unsupported until needed
```
@elBoberido Unfortunately, I don't know the history of this interface.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_response.cpp:
```diff
@@ -17,38 +18,178 @@
+            &iceoryx_client_abstraction->type_supports_,
+            ros_response);
+        }
+        iceoryx_client->releaseResponse(iceoryx_response_payload);
```
@elBoberido Good catch, fixed.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_response.cpp:
```diff
@@ -17,38 +18,178 @@
+    return ret;
+  }
+
+  auto * iceoryx_request_header = iox::popo::RequestHeader::fromPayload(
+    iceoryx_server_abstraction->request_payload_);
```
@elBoberido You are right. IMO this should be fixed in rmw, [see the thread above](https://github.com/ros2/rmw_iceoryx/pull/98#discussion_r1328444615).

elBoberido replied:
Do you know how it is done in other RMW implementations or do they have the same limitation?

mossmaurice replied:
@elBoberido Yes, I think so. Good idea. Anyway, something for a follow-up PR and I'd like to discuss this issue with the `rmw` maintainers first.

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_service_server_is_available.cpp:
```diff
@@ -29,7 +32,29 @@ rmw_service_server_is_available(
+    return ret;
+  }
+
+  std::this_thread::sleep_for(std::chrono::seconds(1));
```
One second seemed like a good magic number to me. I can change this to 42 seconds, though :grin:

elBoberido replied:
Hahaha, in this case I vote for 73 ;)

Joking aside, why is the sleep necessary?

mossmaurice replied:
For the discovery loop to have completed its run. As far as I know, this method is called for the creation of the ROS graph and I also remember there were some workarounds in ROS 2 needed for some buggy behaviour. As I'm a chicken I'll leave it as 1s for now.

See here for the explanation of the method: https://docs.ros.org/en/humble/p/rmw/generated/function_rmw_8h_1af0cec9687e5e4738aba7e061014eb3d5.html

elBoberido replied:
Oh, in this case you could evaluate whether the `iox::roudi::DISCOVERY_INTERVAL` from `iceoryx_posh_types.hpp` would work. In theory this should be the case ... maybe times two to satisfy the inner chicken :)

Oh, I just had an idea. In some examples we just defined a timeout and periodically checked the connection state until either the timeout expired or the expecdted state was reached. Similar to this code (may not compile and you might fix some `!` or `&&`)
```cpp
auto timeout = iox::deadline_timer(n * iox::roudi::DISCOVERY_INTERVAL);
*is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
while(!timeout.hasExpired() && !*is_available) {
    std::this_thread::sleep_for(iox::roudi::DISCOVERY_INTERVAL);
    *is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
}
```
This will massively speed up anything which uses this code, especially when the server is already available. Imagine someone creates multiple clients or maybe it is just a precondition in a loop before the client sends a request, then the unconditional sleep would make the call unbearable slow.

The documentation of `rmw_service_server_is_available` also does not suggest that something will artificially delay this call so it might be surprising for the user. The more I think about it, the more I would even fully remove the sleep since the call states that only the services discovered so far by the local node will be considered.

elBoberido replied:
Oh, it just occurred to me that the proposal with the while loop might not work in cases the user expects the service to not be available. Not sure in which situation this is used generally.

But I have another request. Independent of my proposal above, could you please just use
```cpp
*is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
```
instead of the if-statement?
According to the documentation `is_available` is either `true` or `false` but currently it is only set to `true` in the if-statement. Now let's assume you have a user with this code, which is valid according the documentation
```cpp
boo is_available;
if(rmw_service_server_is_available(node_ptr, client_ptr, &is_available) == RMW_RET_OK && is_available) {
    // do something
}
```
Now some undefined behavior is introduced since `is_available` might not be set. If the compiler inlines `rmw_service_server_is_available` the whole if-statement could be removed and `is_available` be set unconditionally to `true` since not entering the if statement would cause undefined behaviour and since the compiler can assume the code does not execute undefined behaviour, the if-statement must always be `true` ... I guess there is a reason I don't like out-parameter :D

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_service_server_is_available.cpp:
```diff
@@ -29,7 +32,29 @@ rmw_service_server_is_available(
+
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  if (iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED) {
+    *is_available = true;
+  }
+
+  ret = RMW_RET_OK;
+  return ret;
```
@elBoberido No, the return value here notifies that the call was successful aka "Could `is_available` be set correctly". Hence, `is_available = false` and `RMW_RET_OK` is perfectly valid.

elBoberido replied:
I guess I kind of don't like out-parameters :sweat_smile:

mossmaurice commented on rmw_iceoryx_cpp/src/rmw_service.cpp:
```diff
@@ -34,26 +39,94 @@ rmw_create_service(
+    cleanupAfterError();
+    return nullptr;
   }
+  memcpy(const_cast<char *>(rmw_service->service_name), service_name, strlen(service_name) + 1);
```
@elBoberido Because I'm a plumber and not a porcelain guy :clown_face: Kidding aside, this was just a copy & paste. Can be adapted with a bigger refactoring I added it to #99

mossmaurice commented:
@elBoberido This is just a forward port, the changes were already reviewed in #84. Nevertheless, thanks a bunch for your findings :pray: I addressed your comments. Back to you.

elBoberido reviewed (commented):
Looks good. Like I said before, I cannot tell anything about the ROS specific changes.
